src/ordered_set_lang.ml

```diff
@@ -107,128 +107,102 @@ let is_standard t =
     :  t
-    -> parse:(loc:Loc.t -> string -> value)
-    -> standard:value map
-    -> value map
+    -> parse:(loc:Loc.t -> string -> 'a)
+    -> key:('a -> Key.t)
+    -> standard:(Loc.t * 'a) Key.Map.t
+    -> (Loc.t * 'a) Key.Map.t
 end
 
-module Make(Key : Key)(Value : Value with type key = Key.t) = struct
-  module type Named_values = sig
-    type t
-
-    val singleton : Value.t -> t
-    val union : t list -> t
-    val diff : t -> t -> t
-  end
-
-  module Make(M : Named_values) = struct
-    let eval t ~parse ~standard =
-      let rec of_ast (t : ast_expanded) =
-        let open Ast in
-        match t with
-        | Element (loc, s) ->
-          let x = parse ~loc s in
-          M.singleton x
-        | Standard -> standard
-        | Union elts -> M.union (List.map elts ~f:of_ast)
-        | Diff (left, right) ->
-          let left  = of_ast left  in
-          let right = of_ast right in
-          M.diff left right
-      in
-      of_ast t.ast
-  end
-
-  module Ordered = Make(struct
-      type t = Value.t list
-
-      let singleton x = [x]
-      let union = List.flatten
-      let diff a b =
-        List.filter a ~f:(fun x ->
-          List.for_all b ~f:(fun y ->
-            Ordering.neq (Key.compare (Value.key x) (Value.key y))))
-    end)
-
-  module Unordered = Make(struct
-      type t = Value.t Key.Map.t
-
-      let singleton x = Key.Map.singleton (Value.key x) x
-
-      let union l =
-        List.fold_left l ~init:Key.Map.empty ~f:(fun acc t ->
-          Key.Map.merge acc t ~f:(fun _name x y ->
-            match x, y with
-            | Some x, _ | _, Some x -> Some x
-            | _ -> None))
-
-      let diff a b =
-        Key.Map.merge a b ~f:(fun _name x y ->
-          match x, y with
-          | Some _, None -> x
-          | _ -> None)
-    end)
-
-  type value = Value.t
-  type 'a map = 'a Key.Map.t
-
-  let eval t ~parse ~standard =
+module Eval = struct
+  let of_ast ~diff ~singleton ~union t ~parse ~standard =
+    let rec loop (t : ast_expanded) =
+      match t with
+      | Element (loc, s) ->
+        let x = parse ~loc s in
+        singleton x
+      | Standard -> standard
+      | Union elts -> union (List.map elts ~f:loop)
+      | Diff (left, right) ->
+        let left  = loop left  in
+        let right = loop right in
+        diff left right
+    in
     if is_standard t then
-      standard (* inline common case *)
+      standard
     else
-      Ordered.eval t ~parse ~standard
+      loop t.ast
+
+  let ordered eq =
+    let singleton = List.singleton in
+    let union = List.flatten in
+    let diff a b =
+      List.filter a ~f:(fun x ->
+        List.for_all b ~f:(fun y ->
+          not (eq x y)))
+    in
+    of_ast ~diff ~singleton ~union
 
-  let eval_unordered t ~parse ~standard =
-    if is_standard t then
-      standard (* inline common case *)
-    else
-      Unordered.eval t ~parse ~standard
+  let unordered ~singleton ~empty ~merge ~key =
```

Comment:
I tried doing this by passing the module directly, but couldn't figure out the right type annotations. If anyone wants to give it a try, then please do.